crypto/evp/evp_lib.c

```diff
@@ -637,14 +645,21 @@ int EVP_CIPHER_get_key_length(const EVP_CIPHER *cipher)
 
 int EVP_CIPHER_CTX_get_key_length(const EVP_CIPHER_CTX *ctx)
 {
-    int ok;
-    size_t v = ctx->key_len;
-    OSSL_PARAM params[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
-
-    params[0] = OSSL_PARAM_construct_size_t(OSSL_CIPHER_PARAM_KEYLEN, &v);
-    ok = evp_do_ciph_ctx_getparams(ctx->cipher, ctx->algctx, params);
-
-    return ok != 0 ? (int)v : EVP_CTRL_RET_UNSUPPORTED;
+    /* Check to see if we've cached this value previously */
+    if (ctx->key_length < 0) {
+        size_t v = ctx->key_len;
+        OSSL_PARAM params[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
+
+        params[0] = OSSL_PARAM_construct_size_t(OSSL_CIPHER_PARAM_KEYLEN, &v);
+        if (!evp_do_ciph_ctx_getparams(ctx->cipher, ctx->algctx, params))
+            return EVP_CTRL_RET_UNSUPPORTED;
```

Comment:
I am also curious how does this implementation work for the legacy ciphers, i.e. engine based? Are we actually missing a fallback here or what?

Reply:
If we are missing a fallback, it's pre-existing.

Reply:
Yes, I know

Reply:
Legacy implementations cache the key length in the `key_len` context structure field.  This is what `v` was initialised with.  For a legacy cipher, `evp_do_ciph_ctx_getparams()` returns unsupported (-1) which then passes the `ok != 0` check and returns `v`.

So the legacy path was correct in a vile round a bout manner.

Reply:
Ah, I was truly confused by the ok != 0 check. This is really hackish way of the legacy fallback, perhaps some comment should be added. Probably not in this PR though.

Reply:
I expect this will go away as part of this PR.

Reply:
It's still ugly, but not quite so opaque.